lib/pf/Authentication/Source/SAMLSource.pm

```diff
@@ -0,0 +1,244 @@
+
+=cut
+
+sub lasso_login {
+    my ($self) = @_;
+    return Lasso::Login->new($self->lasso_server);
+}
+
+=head2 sso_url
+
+Generate the Single-Sign-On URL that points to the Identity Provider
+
+=cut
+
+sub sso_url {
+    my ($self) = @_;
+    my $url;
+    eval {
+        my $lassoLogin = $self->lasso_login;
+
+        $lassoLogin->init_authn_request($self->idp_entity_id, Lasso::Constants::HTTP_METHOD_REDIRECT);
+        $lassoLogin->request->NameIDPolicy->Format(Lasso::Constants::SAML2_NAME_IDENTIFIER_FORMAT_PERSISTENT);
+        $lassoLogin->request->NameIDPolicy->AllowCreate(1);
+        $lassoLogin->request->ForceAuthn(0);
+        $lassoLogin->request->IsPassive(0);
+        $lassoLogin->request->ProtocolBinding(Lasso::Constants::SAML2_METADATA_BINDING_ARTIFACT);
+
+        $lassoLogin->build_authn_request_msg();
+
+        $url = $lassoLogin->msg_url;
+    };
+    if($@){
+        die "Can't create Single-Sign-On URL : ".$@->{message}."\n";
+    }
+    return $url;
+}
+
+=head2 handle_response
+
+Handle the response from the Identity Provider and extract the username out of the assertion
+
+=cut
+
+sub handle_response {
+    my ($self, $response) = @_;
+    
+    my ($result, $msg) = eval {
+        my $lassoLogin = $self->lasso_login;
+
+        $lassoLogin->process_authn_response_msg($response);
+
+        my $rc = $lassoLogin->accept_sso();
+        if($rc){
+            return ($FALSE, "Single Sign-On failed. Code : $rc");
+        }
+
+        my $assertion = $lassoLogin->get_assertion();
+        my @attribute_list = $assertion->AttributeStatement->Attribute;
+        
+        my $username;
+        foreach my $attribute (@attribute_list){
+            if($attribute->Name eq $self->username_attribute){
+                $username = $attribute->AttributeValue->any->content;
+                last;
+            }
+        }
+
+        if($username){
+            return ($username, "Authentication successful with username : $username");
+        }
+        else {
+            return ($FALSE, "Can't find username in SAML response.")
+        }
+    };
+    if($@){
+        return ($FALSE, "Can't validate Identity provider return message : ".$@->{message});
+    }
+
+    return ($result, $msg);
+
+}
+
+=head2 generate_sp_metadata
+
+Generate the metadata for the Service Provider (this server)
+
+=cut
+
+sub generate_sp_metadata {
+    my ($self) = @_;
+    require pf::config;
+    my $cert = read_file($self->sp_cert_path);
+    $cert = join("\n", map { ($_ !~ /^-----BEGIN CERTIFICATE-----/ && $_ !~ /^-----END CERTIFICATE-----/) ? $_ : () } split(/\n/, $cert));
+    my $vars = {
+        hostname => $pf::config::fqdn,
+        sp_entity_id => $self->sp_entity_id,
+        sp_cert => $cert,
+        entity_id => $self->sp_entity_id,
+        protocol => isenabled($pf::config::Config{captive_portal}{secure_redirect}) ? "https" : "http",
+    };
+
+    our $TT_OPTIONS = {ABSOLUTE => 1};
+    our $template = Template->new($TT_OPTIONS);
```

Comment:
To avoid issues with badly formed xml do the following.
our $TT_OPTIONS = {ABSOLUTE => 1, AUTO_FILTER => 'xml' };
our $template = Template::AutoFilter->new($TT_OPTIONS );